analytics_data_api/v0/views/__init__.py

```diff
@@ -97,3 +108,119 @@ def finalize_response(self, request, response, *args, **kwargs):
+        the fields to exclude with `exclude`.
+
+        ids -- The comma-separated list of identifiers for which results are filtered to.
+            For example, 'edX/DemoX/Demo_Course,course-v1:edX+DemoX+Demo_2016'. Default is to
+            return all courses.
+        fields -- The comma-separated fields to return in the response.
+            For example, 'course_id,created'. Default is to return all fields.
+        exclude -- The comma-separated fields to exclude in the response.
+            For example, 'course_id,created'. Default is to not exclude any fields.
+    """
+    ids = None
+    fields = None
+    exclude = None
+    always_exclude = []
+    serializer_class = serializers.ModelSerializer
+    model = models.Model
+    model_id = 'id'
+
+    def get_serializer(self, *args, **kwargs):
+        kwargs.update({
+            'context': self.get_serializer_context(),
+            'fields': self.fields,
+            'exclude': self.exclude
+        })
+        return self.get_serializer_class()(*args, **kwargs)
+
+    def get(self, request, *args, **kwargs):
+        query_params = self.request.query_params
+        self.fields = split_query_argument(query_params.get('fields'))
+        exclude = split_query_argument(query_params.get('exclude'))
+        self.exclude = self.always_exclude + (exclude if exclude else [])
+        self.ids = split_query_argument(query_params.get('ids'))
+        self.verify_ids()
+
+        return super(APIListView, self).get(request, *args, **kwargs)
+
+    def verify_ids(self):
+        pass  # to be implemented in subclasses
+
+    def default_result(self, item_id):
+        """Default result with fields pre-populated to default values."""
+        result = {
+            self.model_id: item_id,
+        }
+        return result
+
+    def get_result_from_model(self, model, base_result=None, field_list=None):
+        field_list = (field_list if field_list else
+                      [f.name for f in self.model._meta.get_fields()])  # pylint: disable=protected-access
+        result = base_result if base_result else {}
+        result.update({field: getattr(model, field) for field in field_list})
+        return result
+
+    def postprocess_result(self, result):
+        """Applies some business logic to final result without access to any data from the original model."""
+        return result
+
+    def group_by_id(self, queryset):
+        """Return results aggregated by a distinct ID."""
+        formatted_results = []
+        for item_id, model_group in groupby(queryset, lambda x: (getattr(x, self.model_id))):
+            result = self.default_result(item_id)
+
+            for model in model_group:
+                result = self.get_result_from_model(model, base_result=result)
+
+            result = self.postprocess_result(result)
+            formatted_results.append(result)
+
+        return formatted_results
+
+    def get_query(self):
```

Comment:
I did try that double star, but the Q function didn't seem to work with that unfortunately.